iio: dac: ad5766: Add span support.

cristiantpop opened the pull request:
Add support for span for AD5766/5767. This will allow to select between
eight available output ranges and calculate, based on it, an offset and
scale.

Signed-off-by: Denis Gheorghescu <denis.gheorghescu@analog.com>
Signed-off-by: Cristian Pop <cristian.pop@analog.com>

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+		return len;
+
+	// We need to delay it because otherwise the second command will
+	// not execute due to the long time that it takes to fully reset.
```
According to the datasheet, the "minimum time between a reset and the subsequent successful write is typically 25 ns.". Did you see something different? 500 ms doesn't look too good.

mirceacaprioru replied:
Linux coding style uses comments like this:
```
/* We need to delay it because otherwise the second command will
 * not execute due to the long time that it takes to fully reset.
 */
```

cristiantpop replied:
Removed delay.

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -295,6 +451,11 @@ static int ad5766_probe(struct spi_device *spi)
 	return devm_iio_device_register(&spi->dev, indio_dev);
 }
 
+static int ad5766_remove(struct spi_device *spi)
```
This doesn't have anything to do with the span.

cristiantpop replied:
Removed `ad5766_remove` unrelated function withe the commit.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -132,21 +252,47 @@ static const struct ad5766_chip_info ad5766_chip_infos[] = {
 };
 
-static void _ad5766_get_span_range(int *min, int *max)
+static void _get_span_range(struct iio_dev *indio_dev, int *min, int *max)
```
it's nice to keep the `_ad5766` prefix is possible;
especially if it was added in the first place;

also, this function would now need to return int;
so that you can `return 0` on success of `an error code if something occurred;

cristiantpop replied:
Add prefix, return type and return value checking.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+	return sprintf(buf, "%s\n", ad5766_span_ranges[st->span_range]);
+}
+
+static ssize_t  ad5766_get_avaliable_id(const char** available_values, const char *crt_value)
```
there is a kernel lib function called `match_string()` [and another one `sysfs_match_string()`] that does exactly this search in an array of strings
you just need to make sure that the array of strings has a NULL at the end;

it's probably a good idea to use `sysfs_match_string()` ; that one compares strings that end with either `\n` or `\0`

cristiantpop replied:
Use `sysfs_match_string` and remove `ad5766_get_avaliable_id`.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
 			   int *val2,
 			   long m);
 
+static ssize_t ad5766_show_span_range(struct device *dev,
```
this looks like `IIO_CHAN_INFO_SCALE`
maybe check how `ad5761.c` implements it;
the logic for the span range looks similar;
i would adapt from that;

commodo replied:
oh right;
`IIO_CHAN_INFO_SCALE` is already implemented;

well, no idea if this is really needed for debugging only;
but it's not likely to get accepted upstream without a good explanation;

dbogdan replied:
Well, ad5761.c can't adjust the range during runtime.
"offset" is added to "raw" prior to scaling by "scale" in order to obtain value in the specified unit. In the ad5766.c case, "range" (specified directly in volts) is adjusting both "offset" and "scale" accordingly. Modifying only one of these two ("offset", "scale") is not possible given the supported ranges, so they can only be read.

commodo replied:
still, ad5761.c is a good idea as code design;
we can have 
```
struct ad5766_range_params {

};
```

then a table of range_params like in ad5766.c that uses the new range params enums;
```
static const struct ad5766_range_params ad5766_range_params[] = {
        [AD5761_VOLTAGE_RANGE_M10V_10V] = {

        },
        [AD5761_VOLTAGE_RANGE_0V_10V] = {

        },
```

[i didn't do a good job of copy+paste+adapt above]

then a question remains: do we need to set the span-range at run-time for AD5766?
[ if yes, it may be interesting to implement at runtime for AD5761 at some point ]

but, from what i can tell from the datasheet of AD5766, page 29, the span-range is configurable via some power-supplies, which means we'd need to add some regulator[s] to this driver [avdd-supply & avss-supply], then read the values during probe, then determine the span-range [from the table], then set it;
AFAICT, the span-range for AD5766 should be read-only as well

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -241,23 +387,33 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
 
 static int ad5766_setup(struct ad5766_state *st)
 {
+	// We need to delay it because otherwise the second command will
```
in kernel you should use C-style comments;
`/* We need to delay it because otherwise the second command will */`

when using block comments, they need to be formatted like this:
```
	/*
	 * We need to delay it because otherwise the second command will
	 * not execute due to the long time that it takes to fully reset.
	 */
```
or something like that;

but depends on the case, some comments aren't always needed if the code is explicit enough;

cristiantpop replied:
Use C-style comments.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -315,6 +476,7 @@ static struct spi_driver ad5766_driver = {
 		.of_match_table = ad5766_dt_match,
 	},
 	.probe = ad5766_probe,
+	.remove	= ad5766_remove,
```
no need to add an empty remove function;
lately we try to not add anymore remove hooks if we can use `devm_` functions;

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -241,23 +387,33 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+	mdelay(500);
+	// set -5V_to_5V range
+	_ad5766_spi_write(st, AD5766_CMD_SPAN_REG, 6);
+	st->span_range = 6;
```
maybe define some enums instead of `6`
see ad5761.c for more examples;

cristiantpop replied:
Define enum for available ranges.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -241,23 +387,33 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
 		return IIO_VAL_INT;
 	}
+
 	return -EINVAL;
```
at some point, some upstream people may complain that this `return -EINVAL` should be in the `default:` switch case
we can add it in this patchset or later;
but it may be an idea to make it a separate patch for us;

cristiantpop replied:
Move `return -EINVAL` in `default` tag of the switch.

mirceacaprioru commented on drivers/iio/dac/ad5766.c:
```diff
@@ -132,21 +252,47 @@ static const struct ad5766_chip_info ad5766_chip_infos[] = {
 
-	return spi_write(st->spi, &st->data[0].b8[0], 3);
+	for (i = 0; i < strlen(buf); i++) {
+		// since we know the input, we don't need to verify the
```
please change the comments here also

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -88,19 +90,43 @@ struct ad5766_chip_info {
 };
 
+static const char* ad5766_span_ranges[] = {
+	"-20V_to_0V",
```
Doesn't checkpatch complain here? I think it prefers `static const char *ad...`

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+	mdelay(500);
+	
+	if (_ad5766_spi_write(st, AD5766_CMD_SPAN_REG, span_range))
+		return len;
```
nit: add blank line...

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+	for (i = 0; i < ARRAY_SIZE(ad5766_span_ranges); i++)
+	{
+		if (i > 0)
+			ret += sprintf(buf + ret, " ");
```
nit: Add blank line here...

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+		if (i > 0)
+			ret += sprintf(buf + ret, " ");
+		ret += sprintf(buf + ret, "%s", ad5766_span_ranges[i]);
+	}
```
nit: Add blank line here...

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +139,103 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+	}
+
+	mdelay(500);
+	
```
this code does not seem to run in atomic context... Use `msleep()`

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -241,23 +387,33 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+	// not execute due to the long time that it takes to fully reset.
+
+	_ad5766_spi_write(st, AD5766_CMD_SW_FULL_RESET, AD5766_FULL_RESET_CODE);
+	mdelay(500);
```
`msleep()...`

nunojsa commented on drivers/iio/dac/ad5766.c:
```diff
@@ -315,6 +476,7 @@ static struct spi_driver ad5766_driver = {
 	.probe = ad5766_probe,
+	.remove	= ad5766_remove,
 	.id_table = ad5766_spi_ids,
 };
```
Do we need an `id_table`?

cristiantpop commented:
V2:

- Removed delay

- Removed `ad5766_remove` unrelated function withe the commit.

- Add prefix, return type and return value checking.

- Use `sysfs_match_string` and remove `ad5766_get_avaliable_id`.

- Use C-style comments

- Define enum for available ranges.

- Move `return -EINVAL` in `default` tag of the switch.

- Add new lines where necessary.**

dbogdan commented on drivers/iio/dac/ad5766.c:
```diff
@@ -241,23 +389,33 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+		ret = _ad5766_get_span_range(indio_dev, &min, &max);
+		if (ret)
+			return ret;
 		*val = min;
```
min doesn't seem to be a raw value - if this is the case, this is not good ("offset" is added to "raw" prior to scaling by "scale").

buha commented on drivers/iio/dac/ad5766.c:
```diff
@@ -88,19 +89,56 @@ struct ad5766_chip_info {
 	struct spi_device		*spi;
 	struct mutex			lock;
 	const struct ad5766_chip_info 	*chip_info;
+	enum ad5766_voltage_range	span_range;
```
isn't `span_range` something that you want to specify for every channel ? is the range in this chip the same across all channels ?

commodo replied:
looking at page 29 of the datasheet, this looks like a chip-level configuration 
it depends on 2 external voltages

commodo commented:
i know how this sounds, but you can remove this line from the commit `Signed-off-by: Denis Gheorghescu denis.gheorghescu@analog.com`

technically, this work doesn't originate from that author enough to need that Signed-off-by tag

commodo commented:
> technically, this work doesn't originate from that author enough to need that Signed-off-by tag

or does it?
to be honest i did not check the old code;

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -88,19 +89,56 @@ struct ad5766_chip_info {
  */
 
+
+enum ad5766_voltage_range {
```
this enum is being added between the doc-string of `struct ad5766_state` [above] and the struct itself [below];
maybe move it outside of this scope

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -113,9 +151,91 @@ static int ad5766_read_raw(struct iio_dev *indio_dev,
+		       ad5766_set_span_range,
+		       0);
+
+static IIO_DEVICE_ATTR(span_range_available,
```
these 2 `IIO_DEVICE_ATTR` should be dropped;
they are non-standard IIO attributes;
we can probably find ways to implement span-range without these;

`span_range_available` can be implemented with a read_avail hook;
something iike:
```

static int ad7793_read_avail(struct iio_dev *indio_dev,
                             struct iio_chan_spec const *chan,
                             const int **vals, int *type, int *length,
                             long mask)
{
        struct ad7793_state *st = iio_priv(indio_dev);

        switch (mask) {
        case IIO_CHAN_INFO_SCALE:
                *vals = (int *)ad5766_range_params;
                *type = IIO_VAL_INT_PLUS_MICRO;
                *length = ARRAY_SIZE(ad5766_range_params) * 2;

                return IIO_AVAIL_RANGE;
        default:
                return -EINVAL;
        }
}
```

hmm, in this case, we might need to implement `ad5766_range_params` as a 2D matrix;
i am not yet sure about the output at the sysfs; read_avail hooks aren't used as much as they should be;
depending the case, it should be either `IIO_AVAIL_RANGE` or `IIO_AVAIL_LIST`;
still not sure;

cristiantpop replied:
ad5766 has 8 possible output voltage ranges. The selected range, is a common to all channels. So it seems to me it needs to be a global attribute not a channel attribute.
Also It is more logical to me to set a range, and based on it to determine scale and offset.
Than selecting a scale and an offset, for a channel, that will influence the scales and offsets for the other channels. I think that it can be done in this way too. I just have to decide.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -88,19 +89,56 @@ struct ad5766_chip_info {
 	} data[3] ____cacheline_aligned;
 };
 
+static const char *ad5766_span_ranges[] = {
```
this array of string will probably get some questions upstream;
it looks a bit non-standard, so it's hard to explain;

if possible we should try to find ways to remove it and use more of the IIO framework standard stuff;

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -10,6 +10,7 @@
 #include <linux/module.h>
 #include <linux/spi/spi.h>
 #include <linux/iio/iio.h>
+#include <linux/iio/sysfs.h>
```
this include shouldn't be needed

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -132,21 +252,49 @@ static const struct ad5766_chip_info ad5766_chip_infos[] = {
+	buf = ad5766_span_ranges[st->span_range];
+	start = 0;
+
+	for (i = 0; i < strlen(buf); i++) {
```
doing string manipulation like this in the kernel is a bad idea in general;
there are rare occasions that this needs to be done;
and is hard to explain it upstream

cristiantpop commented:
V3:

- Redesign span selection. This will allow to select between the available output ranges trough offset and scale attributes.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+	int i, scale_idx, ret;
+
+	for (i = 0; i < st->crt_scales_avail; i++) {
+		if (st->crt_scale_avail[i][0] == val &&
```
you can remove 1 level of indentation if you do something like:
```
if (st->crt_scale_avail[i][0] != val ||
    st->crt_scale_avail[i][1] != val2)
        continue;
```

commodo replied:
since `st->crt_scale_avail` is pretty long, you can probably assign this to a local variable in this function;
`tbl = st->crt_scale_avail`
then all checks are shorter and fit neater on a single line; lilke `if (tbl[i][0] != val || tbl[i][1] != val2)`

cristiantpop replied:
will do that

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+	for (i = 0; i < st->crt_scales_avail; i++) {
+		if (st->crt_scale_avail[i][0] == val &&
+			st->crt_scale_avail[i][1] == val2) {
+			ret = ad5766_get_scale_idx(st, val, val2, i,
```
isn't `scale_idx == i` here?
if so, do we need th big loop then?

cristiantpop replied:
`scale_idx` and `i` are not always the same.
`crt_scale_avail` table can contain only one valid scale f(offset), `but scale_avail` will always contain `AD5766_VOLTAGE_RANGE_MAX` elements. 
`scale_idx` corresponds to the range that has to be selected.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+			if (ret < 0)
+				return ret;
+
+			ret = _ad5766_spi_write(st, AD5766_CMD_SW_FULL_RESET,
```
why do we need to reset the chip on scale set?
that pretty much resets other registers that are configured?

cristiantpop replied:
This is extracted from the ad5766 datasheet: "Always issue a software reset before writing to the span register."

commodo replied:
hmm;
maybe we should ask upstream if [for this particular driver] we should make the span ranges selectable at runtime;
this could a be a different case than the other chip that @buha is working on;

typically, it's not always a good idea to do reset of a device at runtime;

maybe it's a good idea to ask upstream about this; who knows?

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+	for (i = 0; i < AD5766_VOLTAGE_RANGE_MAX; i++) {
+		if (st->offset_avail[i][0] == val &&
+			st->offset_avail[i][1] == val2) {
+			ret = _ad5766_spi_write(st, AD5766_CMD_SW_FULL_RESET,
```
why do we need to reset the chip on offset set?
that resets other regs

cristiantpop replied:
This is extracted from the ad5766 datasheet: "Always issue a software reset before writing to the span register."

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+				return ret;
+
+			st->crt_range = i;
+			ret = ad5766_set_scale_avail(st);
```
you don't typically need to set the scale available after initialized on probe;
so this can be removed?

cristiantpop replied:
A new table of scales is used after an offset change. Not all scales are compatible with any offset.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+				return ret;
+
+			st->crt_range = scale_idx;
+			ret = ad5766_set_scale_avail(st);
```
you don't typically need to set the scale available after initialized on probe;
so this can be removed?

cristiantpop replied:
This can be removed. I will do that.

commodo commented on drivers/iio/dac/ad5766.c:
```diff
@@ -161,12 +226,165 @@ static int ad5766_write(struct iio_dev *indio_dev, u8 dac, u16 data)
+	int i, ret;
+
+	for (i = 0; i < AD5766_VOLTAGE_RANGE_MAX; i++) {
+		if (st->offset_avail[i][0] == val &&
```
maybe use a temporary variable to make this line shorter?
`tbl = st->offset_avail`

cristiantpop commented:
V4:

- Make a void return for function `ad5766_set_offset_avail` and remove type check
- Make a void return for function `ad5766_calc_scales` and remove type check
- Remove extra line
- Use local variable pointers for shorter lines
- Remove unnecessary `ad5766_set_scale_avail` call in `ad5766_set_scale` function since `crt_scale_avail` wont change here

mirceacaprioru commented:
why was this closed ?
